servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinSelector.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright © 2018-2023 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+import io.servicetalk.client.api.LoadBalancedConnection;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.context.api.ContextMap;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.Predicate;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.Single.failed;
+import static io.servicetalk.concurrent.api.Single.succeeded;
+
+final class RoundRobinSelector<ResolvedAddress, C extends LoadBalancedConnection>
+        implements HostSelector<ResolvedAddress, C> {
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<RoundRobinSelector> indexUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(RoundRobinSelector.class, "index");
+
+    private final String targetResource;
+    private volatile List<Host<ResolvedAddress, C>> hosts;
+    @SuppressWarnings("unused")
+    private volatile int index;
+
+    RoundRobinSelector(String targetResource) {
```

Comment:
nit: for consistency across the repo, consider using `final` where possible, and validate input.